src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+                    continue;
+                }
+
+                ValidateAwsCredentials(awsCredentials);
+
+                if (awsCredentials.Username == null)
+                {
+                    continue;
+                }
+
+                var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+                return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, dateTimeProvider);
+            }
+
+            throw new ArgumentException("A MONGODB-AWS must have access key ID.");
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = securePassword,
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
+        {
+            var username = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var password = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEcsResponse(string relativeUri)
+        {
+            if (relativeUri == null)
+            {
+                return null;
+            }
+
+            var response = AwsHttpClientHelper.GetECSResponse(relativeUri).GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2Response().GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
```

Comment:
As currently written this method never returns null, but the caller checks for a null result.

Either add logic to return null if this method can't create a fully valid AwsCredential or remove the null checking logic in the caller.

Reply:
Should we add an exception in case if only one or two required fields were received?